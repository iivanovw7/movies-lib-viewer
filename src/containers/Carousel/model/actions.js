/**
 * Module contains carousel component
 * @module containers/Carousel/model/actions
 * @author Igor Ivanov
 */

import {
  CAROUSEL_JUMP,
  CAROUSEL_NEXT,
  CAROUSEL_PREV,
  CAROUSEL_DRAG,
  CAROUSEL_DONE,
  CAROUSEL_END,
} from './constants';

/**
 * Move to the next element of carousel.
 * @param {number} length
 *    length value to move.
 * @param {number} skip
 *    number of slides to skip.
 * @return {{length: number, type: string}}
 *    action type and length.
 */
export function carouselNext(length, skip) {
  return {
    type: CAROUSEL_NEXT,
    length,
    skip,
  };
}

/**
 * Move to the desired element of carousel.
 * @param {number} desired
 *    desired number.
 * @return {{type: string, desired: number}}
 *    action type and length.
 */
export function carouselDesired(desired) {
  return {
    type: CAROUSEL_JUMP,
    desired,
  };
}

/**
 * Move to the previous element of carousel.
 * @param {number} length
 *    length value to move.
 * @param {number} skip
 *    number of slides to skip.
 * @return {{length: number, type: string}}
 *    action type and length.
 */
export function carouselPrev(length, skip) {
  return {
    type: CAROUSEL_PREV,
    length,
    skip,
  };
}

/**
 * Carousel dragging action.
 * @param {number} length
 *    length value to move.
 * @return {{length: number, type: string}}
 *    action type and length.
 */
export function carouselDrag(length) {
  return {
    type: CAROUSEL_DRAG,
    length,
  };
}

/**
 * End of carousel position changes.
 * @return {{type: string}}
 *    action type.
 */
export function carouselDone() {
  return {
    type: CAROUSEL_DONE,
  };
}

/**
 * If last slide is displayed
 * @return {{type: string}}
 *    action type.
 */
export function carouselEnd() {
  return {
    type: CAROUSEL_END,
  };
}
