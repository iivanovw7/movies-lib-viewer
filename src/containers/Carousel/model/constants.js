/**
 * Module contains carousel constants
 * @module containers/Carousel/model/constants
 * @author Igor Ivanov
 */
export const CAROUSEL_JUMP = 'containers/Carousel/model/constants/CAROUSEL_JUMP';
export const CAROUSEL_NEXT = 'containers/Carousel/model/constants/CAROUSEL_NEXT';
export const CAROUSEL_PREV = 'containers/Carousel/model/constants/CAROUSEL_PREV';
export const CAROUSEL_DRAG = 'containers/Carousel/model/constants/CAROUSEL_DRAG';
export const CAROUSEL_DONE = 'containers/Carousel/model/constants/CAROUSEL_DONE';
export const CAROUSEL_END = 'containers/Carousel/model/constants/CAROUSEL_END';
