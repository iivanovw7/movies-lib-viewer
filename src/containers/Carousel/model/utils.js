/**
 * Module contains carousel page utils
 * @module containers/Carousel/model/utils
 */

import setting from '../../../config/data';

/**
 * Calculates previous position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} previous slide position.
 */
export function previous(length, current, skip = 0) {
  return Number((current - 1 - skip + length) % length);
}

/**
 * Calculates next position.
 * @param {number} length - slides length.
 * @param {number} current - current slide.
 * @param {number} [skip = 0] - slide to be skipped.
 * @return {number} next slide position.
 */
export function next(length, current, skip = 0) {
  return Number((current + 1 + skip) % length);
}

/**
 * Returns carousel element threshold.
 * @param {Object} target - target element.
 * @return {number} target threshold
 */
export function threshold(target) {
  const width = target.clientWidth;
  return width / 3;
}

/**
 * Carousel swipe event handler.
 * @param {Object} eventData data.
 * @param {number} length - slides length.
 * @param {number} dir - swipe direction.
 * @return {null | number} swipe length or null.
 */
export function swiped(eventData, length, dir = -1) {
  const t = threshold(eventData.event.target);
  const d = dir * eventData.deltaX;

  return d >= t ? length : null;
}

const basePosterUrl = (size) => `${setting.net.tmdbImageUrl}/t/p/w${size}/`;

export const posterUrl = (slide, size) => `${basePosterUrl(size)}${slide.poster_path}`;
