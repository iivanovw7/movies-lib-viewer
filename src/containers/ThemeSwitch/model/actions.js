/**
 * Module contains theme actions
 * @module containers/ThemeSwitch/model/actions
 * @author Igor Ivanov
 */
import { CHANGE_THEME } from './constants';

/**
 * changeTheme action description.
 * @typedef {object} module:containers/ThemeSwitch/model/actions~changeThemeDescription
 * @property {string} type - string action type.
 * @property {string} mode - current mode.
 */
/* eslint-disable import/prefer-default-export */
/**
 * Change theme mode action.
 * @param {string} mode - new theme mode
 *  action locale @see {@link module:containers/ThemeSwitch/model/constants}
 * @return {module:containers/ThemeSwitch/model/actions~changeThemeDescription}
 *  returns action type and payload.
 */
export function changeTheme(mode) {
  return {
    type: CHANGE_THEME,
    mode,
  };
}
/* eslint-enable import/prefer-default-export */
