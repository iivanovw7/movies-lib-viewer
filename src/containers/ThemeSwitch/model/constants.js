/**
 * Module contains theme constants
 * @module containers/ThemeSwitch/model/constants
 * @author Igor Ivanov
 */

import Dark from '../../../../assets/svg/moon.svg';
import Light from '../../../../assets/svg/sun.svg';
import { DARK_THEME, LIGHT_THEME } from '../../../config/constants';
import setting from '../../../config/data';

/**
 * Themes list element description.
 * @typedef {object} module:containers/ThemeSwitch/model/constants~themeListElement
 * @property {String} image - theme image component.
 * @property {String} alt - theme label or alt text.
 */

/**
 * Object contains locales with corresponding images.
 * @constant {Object}
 * @see {@link module:containers/ThemeSwitch/model/constants~themeListElement}
 */
export const THEME_RESOURCES = {
  [DARK_THEME]: {
    alt: 'Dark theme',
    image: Dark,
  },
  [LIGHT_THEME]: {
    alt: 'Light theme',
    image: Light,
  },
};

/**
 * Default theme.
 * @constant {string}
 */
export const DEFAULT_THEME = setting.theme;

/**
 * Change theme action type.
 * @constant {string}
 */
export const CHANGE_THEME = 'containers/ThemeSwitch/model/constants/CHANGE_THEME';
