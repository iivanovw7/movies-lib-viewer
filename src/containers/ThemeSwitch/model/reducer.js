/**
 * Module contains theme reducer
 * @module containers/ThemeSwitch/model/reducer
 * @author Igor Ivanov
 */
import produce from 'immer';

import { CHANGE_THEME } from './constants';

/**
 * Theme reducer initialState.
 * @typedef {object} module:containers/ThemeSwitch/model/reducer~initialState
 * @property {String} mode - current theme mode.
 */

/**
 * Contains initial state.
 * @type {module:containers/ThemeSwitch/model/reducer~initialState}
 */
export const initialState = {
  mode: 'dark',
};

/* eslint-disable default-case, no-param-reassign */
const themeReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case CHANGE_THEME:
        draft.mode = action.mode;
        break;
    }
  });
/* eslint-enable default-case, no-param-reassign */

export default themeReducer;
