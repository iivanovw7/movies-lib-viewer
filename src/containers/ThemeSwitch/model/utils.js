/**
 * Module contains utils used in Theme Switch component
 * @module containers/ThemeSwitch/model/utils
 * @author Igor Ivanov
 */
import { lt, indexOf, inc, always, nth, ifElse, head, pipe, dec, __, length, keys, either } from 'ramda';

import { THEME_RESOURCES, DEFAULT_THEME } from './constants';

/**
 *  Gets string representations of keys out of theme object
 *  @return {Array.<string>} theme names list
 */
const themes = keys(THEME_RESOURCES);
/**
 * Gets index of current locale.
 * @type {f1}
 */
const themeIndex = indexOf(__, themes);
/**
 * Gets length of theme list.
 */
const themesLength = length(themes);
/**
 * Verifies if there is next theme available.
 * @type {f1}
 */
const hasNextTheme = lt(__, dec(themesLength));
/**
 * Returns firs theme, in case if no further themes available.
 */
const firstTheme = head(themes);

/**
 * Function selects next theme name.
 * Returns theme with next index if there is available one or first theme otherwise.
 * @type {Function|*}
 */
const newThemeKey = pipe(themeIndex, ifElse(hasNextTheme, pipe(inc, nth(__, themes)), always(firstTheme)));

/**
 * Returns either a new theme key or default key.
 * @type {Function|*}
 */
const nextThemeKey = either(newThemeKey, DEFAULT_THEME);

export default nextThemeKey;
