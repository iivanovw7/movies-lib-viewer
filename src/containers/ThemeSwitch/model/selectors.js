/**
 * Module contains selectors related to theme .
 * @module containers/ThemeSwitch/model/selectors
 * @author Igor Ivanov
 */
import { createSelector } from 'reselect';

import { initialState } from './reducer';

/**
 * Selector to the theme domain.
 * @method
 * @param {Object} state - application state.
 * @return {string} theme - application theme selector.
 */
const selectTheme = (state) => state.theme || initialState;

/**
 * Select application theme mode.
 * @method
 * @return {Function} creates new theme selector.
 */
const makeSelectTheme = () => createSelector(selectTheme, (themeState) => themeState.mode);

export { selectTheme, makeSelectTheme };
