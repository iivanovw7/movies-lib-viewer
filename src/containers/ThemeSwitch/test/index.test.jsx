/**
 * Module contains ThemeSwitch component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';
import { connect, Provider } from 'react-redux';

import { noop } from '../../../__utils__/stub';
import { storeFactory, testName, history } from '../../../__utils__/utils';
import Button from '../../../components/Button';
import Img from '../../../components/Img';
import { DARK_THEME, LIGHT_THEME } from '../../../config/constants';
import settings from '../../../config/data';
import { OriginalThemeSwitch, mapStateToProps, mapDispatchToProps } from '../index';
import { changeTheme } from '../model/actions';
import { CHANGE_THEME } from '../model/constants';

const baseProps = {
  mode: 'dark',
  size: 55,
  color: 'red',
  style: {},
  onThemeChange: noop,
  history,
};
const stateMock = {
  theme: {
    mode: settings.theme,
  },
};
const store = storeFactory(stateMock);

/* eslint-disable require-jsdoc, react/jsx-props-no-spreading */
// prettier-ignore
describe(testName('ThemeSwitch', 'component test suite'), function ThemeSwitchTestSuite() {
  const Composition = (props, shouldMount = false) => {
    if (shouldMount) {
      const Component = connect(null, { changeTheme })(OriginalThemeSwitch);
      const node = (
        <Provider store={store}>
          <Component {...props} />
        </Provider>
      );

      return mount(node);
    }

    return shallow(<OriginalThemeSwitch {...props} />);
  };

  it('Should match snapshot', () => {
    const component = Composition({...baseProps});

    expect(component)
      .toMatchSnapshot();
  });

  it('Should shallow render <Button />', () => {
    const component = Composition({...baseProps});

    expect(component.find(Button))
      .toHaveLength(1);
  });

  it('Should shallow render <Img />', () => {
    const component = Composition({...baseProps});

    expect(component.find(Img))
      .toHaveLength(1);

    expect( component.find(Img).prop('src') )
      .toBe('SvgrURL');
  });

  it('Should render <ThemeSwitch />', () => {
    const onThemeChangeStub = jest.fn();
    const props = {
      ...baseProps,
      onThemeChange: onThemeChangeStub,
    };
    const component = Composition(props, true);

    component.find('button')
      .at(0)
      .simulate('click');

    expect(onThemeChangeStub)
      .toHaveBeenCalledTimes(1);

    expect(onThemeChangeStub.mock.calls[0][0])
      .toBe(LIGHT_THEME);

  });

  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).onThemeChange(LIGHT_THEME);

    expect(dispatch.mock.calls[0][0])
      .toStrictEqual({
        mode: LIGHT_THEME,
        type: CHANGE_THEME
      });

  });

  it('mapStateToProps', () => {
    expect(mapStateToProps(stateMock))
      .toStrictEqual({
        mode: DARK_THEME
      })
  });

});
/* eslint-enable require-jsdoc, react/jsx-props-no-spreading */
