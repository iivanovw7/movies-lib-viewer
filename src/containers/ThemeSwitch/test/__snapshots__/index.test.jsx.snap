// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mThemeSwitch[39m]: component test suite Should match snapshot 1`] = `
<Button
  hidden={false}
  onClick={[Function]}
  styling={
    Array [
      "height:1.2em;margin-right:0.5em;padding:0;width:1.7em;",
    ]
  }
  target="_blank"
  variant="primary"
>
  <Img
    alt="Light theme"
    className=""
    src="SvgrURL"
    styling={
      Array [
        "height:100%;opacity:0.9;width:100%;",
      ]
    }
  />
</Button>
`;
